src/transformers/models/zamba/modeling_zamba.py

```diff
@@ -0,0 +1,1758 @@
+
+            if layer_type == "attention+mamba":
+                if self.gradient_checkpointing and self.training:
+                    layer_outputs = self._gradient_checkpointing_func(
+                        self.block.__call__,
+                        hidden_states,
+                        original_hidden_states,
+                        layer_idx,
+                        causal_mask,
+                        position_ids,
+                        past_key_values,
+                        output_attentions,
+                        use_cache,
+                        cache_position,
+                    )
+                else:
+                    layer_outputs = self.block(
+                        hidden_states,
+                        original_hidden_states=original_hidden_states,
+                        layer_idx=layer_idx,
+                        attention_mask=causal_mask,
+                        position_ids=position_ids,
+                        past_key_value=past_key_values,
+                        output_attentions=output_attentions,
+                        use_cache=use_cache,
+                        cache_position=cache_position,
+                    )
+                transformer_hidden_states = layer_outputs[0]
+                if output_attentions:
+                    if layer_outputs[1] is not None:
+                        all_self_attns += (layer_outputs[1],)
+                if self.gradient_checkpointing and self.training:
+                    transformer_hidden_states = self._gradient_checkpointing_func(
+                        next(linear_layers).__call__,
+                        transformer_hidden_states,
+                    )
+                else:
+                    transformer_hidden_states = next(linear_layers)(
+                        transformer_hidden_states,
+                    )
+            else:
+                transformer_hidden_states = None
+
+            if self.gradient_checkpointing and self.training:
+                layer_outputs = self._gradient_checkpointing_func(
+                    next(mamba_layers).__call__,
+                    hidden_states,
+                    transformer_hidden_states,
+                    attention_mask,
+                    position_ids,
+                    past_key_values,
+                    output_attentions,
+                    use_cache,
+                    cache_position,
+                )
+            else:
+                layer_outputs = next(mamba_layers)(
+                    hidden_states,
+                    transformer_hidden_states=transformer_hidden_states,
+                    attention_mask=attention_mask,
+                    position_ids=position_ids,
+                    past_key_value=past_key_values,
+                    output_attentions=output_attentions,
+                    use_cache=use_cache,
+                    cache_position=cache_position,
+                )
```

Comment:
Ideally, we would only have self.layers, and iterate over them. Each layer would be either:
- AttentionLayer
- HybridLayer -> Attention + Linear + Mamba
The logic is a bit too complicated here

Reply:
Done. We defined self.layers which contains instantiations of Mamba and Hybrid layers. We defined HybridLayers here https://github.com/Zyphra/transformers_zamba/commit/58d8c2dd95e3a8e8b2d59737ca21b8c4fd92739e#diff-0f4d89960530c068a10af906f1958ed46e3e5f2ff937d6be61517478f383b074R948 which wrap Transformer + Linear + Mamba.